docs/configuration.md

```diff
@@ -1250,6 +1256,27 @@ one of the `allow` regular expressions **and** one of the following holds:
 2.  `deny` is set but no URLs within the manifest match any of the `deny` regular
     expressions.
 
+#### `imageindex`
+
+##### `images`
+
+Set `exist` to false to disable validation that images exist when an image
+index manifest is uploaded.
+
+Set `platforms` to configure the set of platforms to validate the existence of.
+If a platform is included in this list and in the images contained within an index,
+the registry will validate that it exists in the registry. The registry will not
+validate the existence of other platform images in the index.
+
+This parameter does not validate that the platforms are included in the index,
+if an image index is missing one or more platform images from this array it may
+still be accepted by the registry.
```

Comment:
What if i want to configure my registry to always enforce that any image that's referenced in the index, exists?

Requiring me to specify all the potential platforms in advance doesn't seem reasonable (what happens if a new platform type is introduced?)

i'm also wondering what is driving the desire for this level of configuration.  Is there a specific use case that is not met by a simple "allow dangling image references/do not allow dangling image references", which requires the api to allow a user to say "do not allow dangling image references for these platforms, but allow dangling image references for these other platforms"?  If there is, a more straightforward api might be something like:

ExpliciltyAllowedDanglingPlatforms (array of platforms)
ExplicitlyDisallowedDanglingPlatforms (array of platforms)
DefaultDisallowDanglingPlatforms (true/false, default true, which preserves observed behavior in most existing registries as i understand it)

the names need work, but as a general idea, anyway.

Reply:
The idea is that someone running a mirror registry for their network restricted (no internet) environment can then say "we're an x86 shop, enforce that please but ignore the rest".

If Exists is false, nothing is validated.

If Exists is true (or unspecified), and platforms is not specified, everything is validated.

If Exists is true (or unspecified), and there are platforms specified, only they will be validated.

I agree its a little unweildy, and it seems I've explained it very badly in the doc!

Reply:
@bparees I haven't changed the code yet, just the doc, but does this feel better? https://github.com/distribution/distribution/pull/3536/commits/e0685b345a030486667cb58ab3ea1f6509006094

I felt like the issue really was attempting to use a boolean when there are 3 options, and it feels much better to me now :)

I'll code it up in a day or two, but wanted the chance for you to take a look first :)

Reply:
it looks a lot better, but i'd consider going even one step further, by providing for a DenyList as well as an AllowList

ValidatePlatformImageExistence: [all, none, ignoreSelected, validateSelected]
SelectedPlatforms[]  

default to "All" which maintains the behavior most registries implement today.  None of course means do no validation.

ignoreSelected means "ignore references to selected platforms, don't check that the image exists for them"

validateSelected means "if a reference to this platform is present in the manifestlist, validate that the refernced image exists.  if the referenced image doesn't exist, fail"

What is still slightly confusing to a user of this api is whether "validateSelected" means "require that all manifestlists include these platforms" or (what it actually means) "require that all manifestlists that claim to reference these platforms, reference valid images for those platforms"

better wording for the enum and field values can probably help there.

Reply:
I can't see a use case for ignore selected. I can see how someone would want to say "this is the hardware I have" but not "this is the hardware I don't have".

It also suffers from the problem of there being unlimited (in theory) possibilities to enumerate.

> What is still slightly confusing to a user of this api is whether "validateSelected" means "require that all manifestlists include these platforms" or (what it actually means) "require that all manifestlists that claim to reference these platforms, reference valid images for those platforms"

I agree here too, but short of a property called ValidatePlatformImageIfItExistsInTheIndex, I feel like it's best described in the doc, which I already had a go at. Did you think the doc was clear enough there?

Reply:
> I can't see a use case for ignore selected. I can see how someone would want to say "this is the hardware I have" but not "this is the hardware I don't have".

[shrug].  it can always be added later if a need is determined.

> Did you think the doc was clear enough there?

yes, it seemed to explain the behavior clearly